### PR TITLE
profile throughput without new threads

### PR DESCRIPTION
**Pull request by grimoire**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily receiving feedbacks. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit tests to ensure the correctness.
3. If the modification has a dependency on downstream projects of a newer version, this PR should be tested with all supported versions of downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: lmdeploy/pytorch/engine/model_agent.py [hunk 1]
```diff
@@ -742,8 +688,9 @@ async def async_forward(self, inputs: ModelInputs, swap_in_map: SwapMap,
         output = self._forward_impl(inputs,
                                     swap_in_map=swap_in_map,
                                     swap_out_map=swap_out_map)
-        await asyncio.get_event_loop().run_in_executor(None,
-                                                       self.stream.synchronize)
+        await asyncio.sleep(0)
```
Comment (lvhan028) [inline lmdeploy/pytorch/engine/model_agent.py:691]: How does this method outperform the previous one?

File: benchmark/profile_throughput.py [hunk 2]
```diff
@@ -157,18 +156,20 @@ def process_request(self, requests, concurrency, temperature, top_p, top_k,
...
         # start threads
+        tasks = []
         for i in range(concurrency):
-            t = Thread(target=self._inference,
-                       args=(req_queue, res_queue, i, temperature, top_p,
-                             top_k, stream_output),
-                       daemon=True)
-            t.start()
-            threads.append(t)
+            task = self._inference(req_queue, res_queue, i, temperature, top_p,
```
Comment (lvhan028) [inline benchmark/profile_throughput.py:165]: `self._inference` doesn't return a value

File: lmdeploy/pytorch/engine/model_agent.py (hunk 1 above)
Comment (grimoire) [reply lmdeploy/pytorch/engine/model_agent.py:691]: `asyncio.sleep` release CPU without creating new thread

File: lmdeploy/pytorch/kernels/cuda/fill_kv_cache.py [hunk 3]
```diff
@@ -394,15 +324,20 @@ def fill_kv_cache(k_states: Tensor,
...
 
     BLOCK = block_size
     BLOCK_H = triton.next_power_of_2(num_heads)
     BLOCK_D = triton.next_power_of_2(head_dim)
     BLOCK_DV = triton.next_power_of_2(head_dim_v)
-    grid = [batch_size, max_num_blocks]
     kernel_meta = get_kernel_meta(k_states)
     if quant_policy == 0:
+        grid = [num_heads, max_num_blocks, batch_size]
+        num_heads = 1
```
Comment (lvhan028) [inline lmdeploy/pytorch/kernels/cuda/fill_kv_cache.py:339]: why set `num_heads` 1?

**Review by lvhan028 — APPROVED** [review]

File: lmdeploy/pytorch/engine/engine.py [hunk 4]
```diff
@@ -741,6 +748,8 @@ def __update_inputs(next_token_ids):
         logger.debug('<ForwardTask>: '
                      f'batch_size={inputs.seq_length.size(0)} '
                      f'num_tokens={inputs.input_ids.size(-1)}')
+        if self.gpu_count == 1:
```
Comment (AllentDan) [inline lmdeploy/pytorch/engine/engine.py:751]: Why do we need this while previously not?

File: lmdeploy/pytorch/kernels/cuda/apply_rotary_pos_emb.py [hunk 5]
```diff
@@ -60,8 +34,8 @@ def apply_rotary_pos_emb_qk_kernel(
     BLOCK_N: tl.constexpr,
 ):
     """apply rotary on key AND query kernel."""
-    seq_block_id = tl.program_id(0)
-    head_id = tl.program_id(1)
+    seq_block_id = tl.program_id(1)
```
Comment (AllentDan) [inline lmdeploy/pytorch/kernels/cuda/apply_rotary_pos_emb.py:37]: What is the benefit of this change?

File: lmdeploy/pytorch/kernels/cuda/fill_kv_cache.py [hunk 6]
```diff
@@ -79,7 +47,7 @@ def _fill_kv_cache_kernel(
     QSeqLens,
     KVSeqLens,
     BlockOffsets,
-    num_heads: tl.constexpr,
+    is_decoding: tl.constexpr,
```
Comment (AllentDan) [inline lmdeploy/pytorch/kernels/cuda/fill_kv_cache.py:50]: Does it mean that the quant kernel of `fill_kv_cache` can also be optimized?

File: lmdeploy/pytorch/engine/engine.py (hunk 4 above)
Comment (grimoire) [reply lmdeploy/pytorch/engine/engine.py:751]: logits process can be streaming with cuda inputs.
distribute container with cuda tensor is slower than CPU tensor.

File: lmdeploy/pytorch/kernels/cuda/apply_rotary_pos_emb.py (hunk 5 above)
Comment (grimoire) [reply lmdeploy/pytorch/kernels/cuda/apply_rotary_pos_emb.py:37]: increate cos/sin cache hit rate.

File: lmdeploy/pytorch/kernels/cuda/fill_kv_cache.py (hunk 6 above)
Comment (grimoire) [reply lmdeploy/pytorch/kernels/cuda/fill_kv_cache.py:50]: quant kernel would be optimized in future.

**Review by AllentDan — APPROVED** [review]
